### PR TITLE
More extensive check os version number

### PR DESCRIPTION
**Pull request by Carreau**
PyPA `packaging` already normalize it, just need to check that normalization===identity.

### CODE REVIEW

**Comment by takluyver** [comment]
I'd rather avoid depending on packaging so long as it's practical, and in this case I think it is. PEP 440 specifies this format:

```
[N!]N(.N)*[{a|b|rc}N][.postN][.devN]
```

So we just need to translate that to a regex and check that it matches `__version__`. I might warn rather than throwing an error, as versions that don't match that are still allowed.

**Comment by Carreau** [comment]
> So we just need to translate that to a regex and check that it matches __version_

[There](https://github.com/Carreau/pep440/blob/master/pep440/__init__.py#L25-L31https://github.com/Carreau/pep440/blob/master/pep440/__init__.py#L25-L31)

I did it this afternoon. It might be a thing to add into the pep itself maybe.

**Comment by Carreau** [comment]
In case you are lazy:

```
pep440re = re.compile('^([1-9]\d*!)?'        # [N!]
                      '([1-9]\d*)'           # N
                      '(.[1-9]\d*)*'         # (.N)*
                      '((a|b|rc)[0-9]\d*)?' # [{a|b|rc}N]
                      '(\.post[1-9]\d*)?'         # [.postN]
                      '(\.dev[1-9]\d*)?$'         # [.devN]
                      )
```

**Comment by takluyver** [comment]
Nice. Spelling PR coming up ;-)

**Comment by Carreau** [comment]
> Nice. Spelling PR coming up ;-)

Feel free to push/release on PyPI.

**Comment by Carreau** [comment]
Removed dependency. Still raising. Do you prefer log-info , log-warn, log-debug, or keep raise ?

File: flit/common.py [hunk 1]
```diff
@@ -62,19 +80,32 @@ def get_info_from_module(target):
...
+    if not isinstance(version, str):
         raise InvalidVersion('__version__ must be a string, not {}.'
-                                .format(type(module_version)))
-    if not module_version[0].isdigit():
+                                .format(type(version)))
+    if not version[0].isdigit():
         raise InvalidVersion('__version__ must start with a number. It is {!r}.'
-                                .format(module_version))
+                                .format(version))
+    if _is_canonical(version):
```
Comment (takluyver) [inline flit/common.py]: I think you're missing a `not` here ;-)

File: flit/common.py [hunk 2]
```diff
@@ -6,6 +6,24 @@
...
+                      '([1-9]\d*)'           # N
+                      '(.[1-9]\d*)*'         # (.N)*
+                      '((a|b|rc)[0-9]\d*)?' # [{a|b|rc}N]
+                      '(\.post[1-9]\d*)?'         # [.postN]
+                      '(\.dev[1-9]\d*)?$'         # [.devN]
+                      )
+
+def _is_canonical(version)->bool:
+    """
+    Retunr wether or not the version string is canonical according to Pep 440
```
Comment (takluyver) [inline flit/common.py]: 'Return whether'

**Comment by takluyver** [comment]
I think a warning would be the appropriate level - it indicates the user is doing something non-optimal but still allowed.

File: flit/common.py (hunk 1 above)
Comment (Carreau) [inline flit/common.py]: `not` not not added.

File: flit/common.py (hunk 2 above)
Comment (Carreau) [inline flit/common.py]: Fixed.

**Comment by Carreau** [comment]
And warning.

**Comment by Carreau** [comment]
Need to fix docstring/test.

**Comment by Carreau** [comment]
I think I need to rethink the regex, as things can be `0|[1-9]\d+` like `dev0` is allowed.

**Comment by Carreau** [comment]
OK, works now.

**Comment by takluyver** [comment]
:+1:

**Pull request merged**